server/src/test/java/org/elasticsearch/index/engine/LuceneChangesSnapshotTests.java

```diff
@@ -330,10 +342,42 @@ public void run() {
     public void testOverFlow() throws Exception {
         long fromSeqNo = randomLongBetween(0, 5);
         long toSeqNo = randomLongBetween(Long.MAX_VALUE - 5, Long.MAX_VALUE);
-        try (Translog.Snapshot snapshot = engine.newChangesSnapshot("test", fromSeqNo, toSeqNo, true, randomBoolean())) {
+        try (Translog.Snapshot snapshot =
+                 engine.newChangesSnapshot("test", fromSeqNo, toSeqNo, true, randomBoolean(), true, randomBoolean())) {
             IllegalStateException error = expectThrows(IllegalStateException.class, () -> drainAll(snapshot));
             assertThat(error.getMessage(),
                 containsString("Not all operations between from_seqno [" + fromSeqNo + "] and to_seqno [" + toSeqNo + "] found"));
         }
     }
+
+    public void testStats() throws Exception {
+        try (Store store = createStore();
+             Engine engine = createEngine(defaultSettings, store, createTempDir(), NoMergePolicy.INSTANCE)) {
+            final int numOps = between(100, 5000);
+            applyOperations(engine, generateHistoryOnReplica(numOps, randomBoolean(), randomBoolean(), randomBoolean()));
+            engine.refresh("test");
+            // Can't access stats if didn't request it
+            try (Translog.Snapshot snapshot = engine.newChangesSnapshot("test", 0, Long.MAX_VALUE, false, randomBoolean(), true, false)) {
+                IllegalStateException error = expectThrows(IllegalStateException.class, snapshot::totalOperations);
+                assertThat(error.getMessage(), equalTo("Access stats of a snapshot created with [access_stats] is false"));
+                final List<Translog.Operation> translogOps = drainAll(snapshot);
+                assertThat(translogOps, hasSize(numOps));
+                error = expectThrows(IllegalStateException.class, snapshot::totalOperations);
+                assertThat(error.getMessage(), equalTo("Access stats of a snapshot created with [access_stats] is false"));
+            }
+            // Access stats and operations
+            try (Translog.Snapshot snapshot = engine.newChangesSnapshot("test", 0, Long.MAX_VALUE, false, randomBoolean(), true, true)) {
+                assertThat(snapshot.totalOperations(), equalTo(numOps));
+                final List<Translog.Operation> translogOps = drainAll(snapshot);
+                assertThat(translogOps, hasSize(numOps));
+                assertThat(snapshot.totalOperations(), equalTo(numOps));
+            }
+            // Access stats only
+            try (Translog.Snapshot snapshot = engine.newChangesSnapshot("test", 0, Long.MAX_VALUE, false, randomBoolean(), false, true)) {
```

Comment:
Can we randomly use max-seq-no from engine instead of `Long.MAX_VALUE`? Also in the other verifications update in this test case.